console/src/view/table.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{
+    input, state,
+    view::{self, bold},
+};
+use std::convert::TryFrom;
+use tui::{
+    layout,
+    text::{self, Span, Spans, Text},
+    widgets::TableState,
+};
+
+use std::cell::RefCell;
+use std::rc::Weak;
+
+pub(crate) trait TableList<R, S: TryFrom<usize>> {
+    fn header(&self) -> &'static [&'static str];
+
+    fn state_mut(&mut self) -> &mut TableListState<R, S>;
+
+    fn state(&self) -> &TableListState<R, S>;
```

Comment:
this...feels kind of strange to me. the fact that a `TableList` is expected to _own_ a `TableListState`, but then it's only ever implemented _for_ `TableListState` seems like a kind of weird factoring. what do you think about changing the trait so that more of its methods are actually methods _on_ `TableListState`, like this:

```rust
pub(crate) trait TableList {
    type Row;
    type Sort: SortBy + TryFrom<usize>;

    fn header() -> &'static [&'static str];

    fn render<B: tui::backend::Backend>(
        state: &mut TableListState<Self>,
        styles: &view::Styles,
        frame: &mut tui::terminal::Frame<B>,
        area: layout::Rect,
        state: &mut state::State,
    );
}

pub(crate) struct TableListState<T: TableList> {
    pub(crate) sorted_items: Vec<Weak<RefCell<T::Row>>>,
    pub(crate) sort_by: T::Sort,
    pub(crate) selected_column: usize,
    pub(crate) sort_descending: bool,
    pub(crate) table_state: TableState,
}

impl<T: TableList> TableListState<T> {
    pub(in crate::view) fn len(&self) -> usize {
        self.sorted_items.len()
    }

    pub(in crate::view) fn update_input(&mut self, event: input::Event) {
         // ...
    }

    pub(in crate::view) fn key_input(&mut self, input::KeyEvent { code, .. }: input::KeyEvent) {
        use input::KeyCode::*;
        let header_len = T::header().len();
        match code {
            Left => {
                if self.selected_column == 0 {
                    self.selected_column = header_len - 1;
                } else {
                    self.selected_column -= 1;
                }
            }
            Right => {
                if self.selected_column == header_len - 1 {
                    self.selected_column = 0;
                } else {
                    self.selected_column += 1;
                }
            }
            Char('i') => {
                self.sort_descending = !self.sort_descending
            }
            Down => self.scroll_next(),
            Up => self.scroll_prev(),
            _ => {} // do nothing for now...
        }
        if let Ok(sort_by) = S::try_from(self.selected_column) {
            self.sort_by = sort_by;
        }
    }

    pub(in crate::view) fn scroll_with(&mut self, f: impl Fn(&Vec<Weak<RefCell<T::Row>>>, usize) -> usize) {
        // If the list of sorted items is empty, don't try to scroll...
        if self.sorted_items.is_empty() {
            self.table_state.select(None);
            return;
        }

        // Increment the currently selected row, or if no row is selected, start
        // at the first row.
        let i = self.table_state.selected().unwrap_or(0);
        let i = f(&self.sorted_items, i);
        self.table_state.select(Some(i));
    }

    pub(in crate::view) fn scroll_next(&mut self) {
         // ...
    }

    pub(in crate::view) fn scroll_prev(&mut self) {
        // ...
    }

    pub(in crate::view) fn controls(&self, styles: &view::Styles) -> Text {
        // ...
    }
    
    pub(in crate::view) fn render<B: tui::backend::Backend>(
        &mut self,
        styles: &view::Styles,
        frame: &mut tui::terminal::Frame<B>,
        area: layout::Rect,
    ) {
        T::render(self, styles, frame, layout, self.state)
    }
}
```
Then, we could implement `TableList` for zero-sized structs that just define the header and `render` function, like this:

```rust
#[derive(Debug, Default)]
struct ResourcesState {}

impl TableList for ResourcesState {
    type Row = Resource;
    type Sort = resources::SortBy;

    fn header(&self) -> &'static [&'static str] {
        &[
            "ID",
            "Kind",
            "Total",
            "Target",
            "Type",
            "Location",
            "Attributes",
        ]
    }

    fn render<B: tui::backend::Backend>(
        state: &mut TableListState<Self>,
        styles: &view::Styles,
        frame: &mut tui::terminal::Frame<B>,
        area: layout::Rect,
        state: &mut state::State,
    ) {
       // ... actually render the table ...
    }
}
```

this would also give us a place to eventually factor out some of the common parts of the `render` function.

what do you think? this might end up being just as complex as the current implementation...